Add HTTP basic auth docs

chrismwendt opened the pull request:
The only wrinkle in using nginx for HTTP basic auth is that the request immediately following the WebSocket request to go-langserver results in a 401 (it happens to be a telemetry request):

![image](https://user-images.githubusercontent.com/1387653/53929894-3739cf00-4044-11e9-8276-e140f7965351.png)

The response is 401 from the Go frontend because I think nginx is also sending the basic auth to non-`/go` routes. I discovered this by noticing that the Network tab did not show basic auth, but the nginx access log did. Subsequent requests to non-`/go` routes do not have the basic auth added.

@keegancsmith Do you know why nginx seemingly attaches the basic auth defined in the `/go` location directive to the `/` location directive for the subsequent request?

keegancsmith commented on README.md:
````diff
@@ -12,13 +12,68 @@ Docker image `sourcegraph/lang-go` from Docker Hub.
+
+```
+  "go.serverUrl": "ws://langserveruser:PASSWORD@host.docker.internal:4389/go",
+  "go.sourcegraphUrl": "http://host.docker.internal:7080",
````
given the config below I would of thought you would of used the same port for serverUrl as sourcegraphUrl.

chrismwendt replied:
Oh you're right, I changed this to 7080 to match the `sourcegraphUrl`.

keegancsmith commented on README.md:
```diff
@@ -12,13 +12,68 @@ Docker image `sourcegraph/lang-go` from Docker Hub.
+
+Finally, restart the sourcegraph/server container to pick up the configuration change.
+
+After deploying the language server, unauthenticated access to `http://localhohst:7080/go` should be blocked, but code intelligence should work when you're logged in.
```
```suggestion
After deploying the language server, unauthenticated access to `http://localhost:7080/go` should be blocked, but code intelligence should work when you're logged in.
```

beyang commented on README.md:
```diff
@@ -12,13 +12,68 @@ Docker image `sourcegraph/lang-go` from Docker Hub.
-Anyone that connects to the language server can access resources such as private
-code that the language server has access to.
+If you have private code, we recommend deploying the language server behind an
+auth proxy (such as HTTP basic authentication below), a firewall, or a VPN.
```
I wouldn't call HTTP basic authentication an "auth proxy". "auth-proxy" is the service that proxies traffic and enforces auth; HTTP basic auth is an auth mechanism, which could be enforced by the service, itself.

```suggestion
auth proxy (such as the example below using HTTP basic authentication in NGINX), a firewall, or a VPN.
```

beyang commented:
The main issue with this approach, in my mind, is that it requires passing the access token to the client and if this token were ever leaked, it would give token bearer access to the language server. However, the same holds true for any PAT or session cookie. The access token is like a PAT/cookie that only expires when the site admin manually rotates it. This is probably acceptable to a wide range of customers, because all the attack vectors for getting the LS access token would also apply to PATs and session cookies; the only difference is the expiry. Is there anything else I'm missing?

The auth proxy impl would give a security guarantee on par with what is provided by the Sourcegraph frontend (i.e., instead of a token that expires when the site admin changes it, you get a token that is expired whenever a user revokes it or one that has an expiry). The v1 of that would probably look something like:
* Add the same global settings fields as in this PR.
* Modify the frontend HTTP handler to attach additional HTTP handlers rooted under subpaths like `/go`. It would create these dynamically given the global settings.
* Everything else would be the same as in this PR.

So not a *huge* amount of work, but non-trivial to write, test, and review well. I'm okay with this PR, given that we are explicit about the security model in the documentation.

chrismwendt commented:
@keegancsmith I added a few more notes for admins who have deployed Sourcegraph elsewhere. This solution does not require any code changes, and it's pretty close to proxy-in-Sourcegraph, so I'm going to merge as-is. We can still do the proxy-in-Sourcegraph later if we want.

> Is there anything else I'm missing?

@beyang Nope, you're right that the basic auth acts like an access token and revoking it is a matter of deleting the old `.htpasswd` file and recreating it with a new password. I added a sentence about revoking the password.
